opt/optbuilder: Add name resolution tests

rytaft opened the pull request:
This commit copies the relevant tests from
`sql/sem/tree/name_resolution_test.go` into the `optbuilder`
to test that `optbuilder/scope` correctly resolves names.

Release note: None

rytaft committed:
sql: Refactor name resolution tests

This commit refactors some of the tests from
sql/sem/tree/name_resolution_test.go that are also relevant to
the optimizer. It copies these tests into testutils/sqlutils, and
adds an interface that is implemented by sql/sem/tree/fakeScope and
optbuilder/scope to test that they correctly implement
tree.ColumnItemResolver.

Release note: None

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LCelxDi5Q_E3gFf2KWy:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Indeed the same question as Andy. You are not testing anything about scoping here. I am :-1: on merging this PR as-is.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LChJkryECv90y_tpEkJ:ble1ken)*
<!-- Sent from Reviewable.io -->

knz commented:
Oh I see it now. What's being tested here is the implementation of the name resolution interfaces by `opt.scope`. Sorry for the earlier knee-jerk reaction.

OK But then still the amount of code duplication here won't do. The problem you're creating is that the core of the name resolution test is the list of test cases and the "backbone" that runs the parser and checks that each name resolution result matches the expectations. That core of the test must exist in a single place as a single source of truth.

Meanwhile I recognize your interest to validate via testing that the `opt.scope` code properly implements the interface. Arguably the lack of test in `sqlbase.ColumnItemResolver` and `sql.SchemaResolver` was a mistake! So I'm glad you're taking the initiative here.

So what do we do from there?

I would personally recommend the following:

- export `opt.scope`
- make `name_resolution_test.go` in `sem/tree` live in package `tree_test` instead of `tree`, so that it can depend on `opt` alongside `tree`. Change the code to prefix the `tree.` package name where appropriate.
- rename the two test functions `TestResolveQualifiedStar`, `TestResolveTablePatternOrName` to `runResolveQualifiedStarTest` and `runResolveTablePatternOrNameTest`, and give them a new argument of type `tree.ColumnItemResolver`
- create new tests `TestResolveQualifiedStar` and `TestResolveTablePatternOrName` that call the `run` functions each two times successively, a first time on the locally defined `fakeSource`, and a second time on the `opt.Scope` instance defined by your code. 

What you achieve this way is twofold: 
1) checking that the name resolution code itself is correct (testing against `fakeSource`) and 
2) that the `opt.Scope` does its job (assuming the name resolution code is correct, tested by the first call).

All using a single test routine, with a healthy dose of DRY. 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LChWFlB7inLAsceIn8V:bdr72cl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I'm not crazy about exporting `opt.scope` just for this test. We could instead export a function (e.g. `CreateTestingColumnItemResolver`) which returns it as a `ColumnItemResolver`.

Though I'm not crazy about the `sem/tree` test running tests against `opt` either.. It could get confusing when something fails in `tree` because of a change that touches `opt`. I would make the test code available in `tree` as a "library" (via an interface that extends `ColumnItemResolver` and adds a few methods to allow adding the test table info).  TestXX code would just call into this little library to run the tests against its implementation of that interface.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LChjDZpCaJTY89YVsJB:bwwa4df)*
<!-- Sent from Reviewable.io -->

knz commented:
I think I agree with you Radu. One possibility is to make the name resolution test live in a separate package, separate from both `tree` and `opt`. Then use a test constructor in `opt` that returns a `ColumnItemResolver` for this test, as you suggest.

rytaft commented:
`tree` seems like a good place for it since `optbuilder` already depends on `tree`. I can just make a separate file (e.g., `name_resolution_testutils.go`) to contain the "library" interface. Does that sound reasonable?

---







*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LCiMlgQ3pG5x3T980HB:bxbx3c4)*
<!-- Sent from Reviewable.io -->

knz commented:
no objections

andy-kimball commented:
I see now what you're trying to test: your implementation of the `tree.ColumnItemResolver` interface (a comment in the test would have been useful). I think it's fine not to DRY up the test code, since the cost of doing that is even tighter binding between packages. That said, I'm also fine with Radu's suggestion to add a testing helper function, as long as it doesn't turn into an elaborate, hard-to-understand abstraction.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LCiPk4C2dQzw0AmSIEu:b-a9ik4k)*
<!-- Sent from Reviewable.io -->

knz commented:
>  I think it's fine not to DRY up the test code

I think that any sentence that starts with these words is suspicious ;) More specifically:

> since the cost of doing that is even tighter binding between packages. 

If that is what concerns you, we can discuss alternate ways to DRY that do not create this tighter binding. For example, one could move the entire name resolution logic (`name_resolution.go`) into its own package!

>  as long as it doesn't turn into an elaborate, hard-to-understand abstraction.

having a constructor for an object that implements the `ColumnItemResolver` interface doesn't look like an "elaborate, hard-to-understand" abstraction. YMMV?

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LCicaj6A8rgJwuQmlMp:b7b5oey)*
<!-- Sent from Reviewable.io -->

rytaft commented:
I think I've implemented all the suggestions. I added the refactored code to `testutils` since adding it to `tree` caused a dependency cycle (but feel free to suggest an alternate location...).

I updated the commit message as well. PTAL.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LDIOLiBD438y6Rbv0WE:b-76jpln)*
<!-- Sent from Reviewable.io -->

rytaft commented:
@knz does the new version look ok to you? @RaduBerinde and @andy-kimball any objections?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LDfSZtV3A2E2nLqJnia:b-gd8oqg)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LDgUdIG6dNSZwO12khd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks for walking the extra mile!

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LDgVMFJCqa1EUsBZzGm:b-y32unt)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks! @andy-kimball I'll wait for #26143

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25588#-:-LDgXPh01XXsgQcSZ9G9:b-u6jood)*
<!-- Sent from Reviewable.io -->
